### ISSUE TITLE
Race condition with interleaved messages using TCS

### ISSUE THREAD

**Issue by yevhen** [dotnet/orleans#4774]
We're extensively using the async batcher pattern described by @hataytuna in this book
https://github.com/hataytuna/Distributed/raw/master/Applied%20Actor%20Model%20with%20Orleans%20v1.1.pdf It is crucial to performance boosts we got from batching in our system.

We figured out that problem we're experiencing with graceful shutdown described in #4757 is due to Orleans not being able to deactivate some of the grains. This only happens when TaskCompletionSource is used (along with the buffering and timer-based completion).

I have a failing test (#4773) for this scenario you can pull and check. On the surface, it looks like after DeactivateOnIdle is called Orleans disposes the timer while there interleaved requests, which leaves tasks in the uncompleted state, thus preventing Orleans from deactivating the grain.

Please, advice what can be done wrt to this issue.

**Comment by yevhen** [comment]
Found bug in my test. Seem that it works. Now I'm completely confused why exactly the same code (and test) doesn't work properly in my app.

**Comment by yevhen** [comment]
Well, we found that problem is with one of out actor middlewares. After we have removed the problematic actor filter it started to work. We'll be investigating the reasons. Sorry for false alarm.

**Comment by hataytuna** [comment]
Glad to hear it.

Cheers,
Hatay

________________________________

**Comment by yevhen** [comment]
Unfortunately, it's still an issue. I've added more repeats to the test so now it fails more reliably. Please confirm if the test is wrong or there is an issue. None of the things helped and we still have some actors stuck in production environment during rolling restart :(

**Comment by yevhen** [comment]
Ok, guys, now I can definitely confirm that it's the bug in Orleans code. 

I've changed [this](https://github.com/dotnet/orleans/blob/master/src/Orleans.Runtime/Catalog/ActivationData.cs#L394) method to:

```cs
        public void PrepareForDeactivation()
        {
            SetState(ActivationState.Deactivating);
            deactivationStartTime = DateTime.UtcNow;
            if (!IsCurrentlyExecuting)
                StopAllTimers();
        }
```
(did custom private build) and now my system shuts down in production smoothly without any issues.

I don't know how to write the test which will fail reliably (to show the presence of the bug) but it can be easily figured out analytically. Take look at this method:

https://github.com/dotnet/orleans/blob/839e95303f897c0bcc6b86222ff03d43b1420e21/src/Orleans.Runtime/Catalog/Catalog.cs#L867-L886

It checks whether the activation is still executing requests and if yes - it subscribes to OnIdle to finish deactivation. But since timers are already stopped by `PrepareForDeactivation` those requests are hanging since nothing is able to complete them - the timer is gone.

I'v checked all paths to this method and it seems that this fix is right, once the activation is done with outstanding requests the OnIdle will be called and timers will be properly disposed anyways. All the other places has an inavariant that no requests are currently executing (ie Activation collection) and so the timers will be stopped immediately.

This bug was in Orleans since the beginning (as blame trace shows). Not sure how anyone was using this pattern (TaskCompletionSource) without issues (cc @hataytuna).

**Comment by benjaminpetit** [comment]
@yevhen Thanks you so much for this investigation. Trying to understand what was the logic behind this.

**Comment by benjaminpetit** [comment]
So I am investigating on this for some times now, and there are things that I don't understand.

> It checks whether the activation is still executing requests and if yes - it subscribes to OnIdle to finish deactivation. But since timers are already stopped by PrepareForDeactivation those requests are hanging since nothing is able to complete them - the timer is gone.

Grain timer tick != Grain request.

When a timer is executing, `ActivationData.IsCurrentlyExecuting` will still return `false`, since the timer does not update the `ActivationData` (if I am not mistaken, single threaded is guarantee with the scheduler). That's why we have a specific method `ActivationData.WaitForAllTimersToFinish`.

The call `StopTimer` from `ActivationData.PrepareForDeactivation` will only prevent new ticks to be scheduled and executed, I don't think it has any impact on a possible currently running tick.

Do you have logic in your grain where some request wait for a reminder to fire?

**Comment by yevhen** [comment]
It's all due to interleaving. When the request for deactivation comes in the grain still has some of the requests incompleted. In this case `ActivationData.IsCurrentlyExecuting` will return `true` and so deactivation will be delayed until idle. But the problem that those interleaved requests can only be completed by the timer which has been stopped already. It might be possible that at the time when deactivation request is issued the timer is about to tick (no timer tick is currently executing) and interleaved requests are already queued (buffered inside the grain).  Stopping the timer on deactivation means no next tick will be delivered and so requests are hanging ...

**Comment by benjaminpetit** [comment]
Ok I think I got it by reading again your failing test. Let me rephrase to see if I understand correctly:

On a given grain, you have multiple pending requests (interleaving) that are waiting a timer tick to complete. If you stop the timer before is has the chance to unblock the pending requests, the grain will never be able to deactivate itself.

I think the issue is more general: if the grain is currently executing a request, and for some reason this request is stuck (deadlock, ect), the deactivation logic on the OnIdle will never be called, so the grain will not be deactivated correctly. I think your fix make sense, but we need to have more generic safeguard to catch this kind of possible issues.

**Comment by yevhen** [comment]
@benjaminpetit thanks for #4830 fix. We also found another problem with deactivation :(

Some of our grains do retries in timer callbacks and have code similar to the one below (simplified):
```cs
while (true)
{
    if (TryFoo())
        break;
    await Task.Delay(15);
}
```

Orleans unable to destroy such actors in time due to this line of code
https://github.com/dotnet/orleans/blob/e2b25a695767ae5746ffe47218c1331427997894/src/Orleans.Runtime/Catalog/Catalog.cs#L963

not having any timeout attached. Basically, the deactivaion procedure hangs while actor is doing something heavy (lengthy) inside timer callback.

In our private Orleans build I added WithTimeout and the simply proceed to force destroy of such actors. This to add to your list of issues of what could go wrong during deactivation stage. You mentioned a general check for actors which stuck while handling requests but this is about actors which stuck during timer callback.
